code/modules/library/library_admin.dm

```diff
@@ -0,0 +1,212 @@
+	var/list/reports = list()
+
+	///TGUI page we are currently on
+	var/page_state = LIBRARY_MENU_MAIN
+	///Ckey's books we are viewing
+	var/selected_ckey
+
+	///information for the book we are opening in browserui
+	var/datum/cachedbook/view_book
+	///browserui helper variable for turning pages in book
+	var/view_book_page = 0
+
+/datum/ui_module/library_manager/ui_interact(mob/user, ui_key = "main", datum/tgui/ui = null, force_open = FALSE, datum/tgui/master_ui = null, datum/ui_state/state = GLOB.admin_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "LibraryManager", name, 600, 600, master_ui, state)
+		ui.autoupdate = TRUE
+		ui.open()
+
+/datum/ui_module/library_manager/ui_data(mob/user)
+	var/list/data = list()
+	data["pagestate"] = page_state
+	data["booklist"] = cached_books
+	data["ckey"] = selected_ckey ? selected_ckey : "ERROR"
+	data["reports"] = reports
+	for(var/CB in cached_books)
+
+	data["modal"] = ui_modal_data(src)
+	return data
+
+/datum/ui_module/library_manager/ui_act(action, params)
+	if(..())
+		return
+
+	if(ui_act_modal(action, params))
+		return
+
+	switch(action)
+		if("view_reported_books")
+			reports= list()
+			for(var/datum/cachedbook/CB in  GLOB.library_catalog.getFlaggedBooks())
+				for(var/r in CB.reports)
+					var/datum/flagged_book/report = r
+					if(!report)
+						return
+					var/datum/library_category/report_category =  GLOB.library_catalog.getReportCategoryByID(report.category_id)
+					var/report_info = list(
+						"reporter_ckey" = report.reporter,
+						"uploader_ckey" = CB.ckey,
+						"id" = CB.id,
+						"title" = CB.title,
+						"author" = CB.author,
+						"report_description" = report_category.description,
+					)
+					reports += list(report_info)
+			page_state = LIBRARY_MENU_REPORTS
+		if("delete_book")
+			if(GLOB.library_catalog.removeBookByID(text2num(params["bookid"])))
+				log_admin("[key_name(usr)] has deleted the book [params["bookid"]].")
+				message_admins("[key_name_admin(usr)] has deleted the book [params["bookid"]].")
+		if("view_book")
+			if(params["bookid"])
+				view_book_by_id(text2num(params["bookid"]))
+
+		if("return")
+			page_state = LIBRARY_MENU_MAIN
+
+
+/datum/ui_module/library_manager/proc/ui_act_modal(action, list/params)
+	. = TRUE
+	var/id = params["id"] // The modal's ID
+	var/list/arguments = istext(params["arguments"]) ? json_decode(params["arguments"]) : params["arguments"]
+	switch(ui_modal_act(src, action, params))
+		if(UI_MODAL_OPEN)
+			switch(id)
+				if("specify_ssid_delete")
+					ui_modal_input(src, id, "Please input a book SSID:", null, arguments)
+				if("specify_ckey_search")
+					ui_modal_input(src, id, "Please input a CKEY:", null, arguments)
+				if("specify_ckey_delete")
+					ui_modal_input(src, id, "Please input a CKEY:", null, arguments)
+				else
+					return FALSE
+		if(UI_MODAL_ANSWER)
+			var/answer = params["answer"]
```

Comment:
Prob want to clean this param up before you send it to other admins. Just in case

Reply:
sanitized this variable where needed better

Reply:
cleaned up